state-chain/chains/src/ccm_checker.rs

```diff
@@ -0,0 +1,317 @@
+use crate::{
+	sol::{
+		SolAsset, SolCcmAccounts, SolPubkey, CCM_BYTES_PER_ACCOUNT, MAX_CCM_BYTES_SOL,
+		MAX_CCM_BYTES_USDC,
+	},
+	CcmChannelMetadata,
+};
+use cf_primitives::{Asset, ForeignChain};
+use codec::{Decode, Encode};
+use scale_info::TypeInfo;
+use sp_std::vec::Vec;
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Encode, Decode, TypeInfo)]
+pub enum CcmValidityError {
+	CannotDecodeCfParameters,
+	CcmIsTooLong,
+	CfParametersContainsInvalidAccounts,
+}
+
+pub trait CcmValidityCheck {
+	fn check_and_decode(
+		_ccm: &CcmChannelMetadata,
+		_egress_asset: cf_primitives::Asset,
+	) -> Result<DecodedCfParameters, CcmValidityError> {
+		Ok(DecodedCfParameters::NotRequired)
+	}
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum DecodedCfParameters {
+	Solana(SolCcmAccounts),
+	NotRequired,
+}
+
+pub struct CcmValidityChecker;
+
+impl CcmValidityCheck for CcmValidityChecker {
+	/// Checks to see if a given CCM is valid. Currently this only applies to Solana chain.
+	/// For Solana Chain: Performs decoding of the `cf_parameter`, and checks the expected length.
+	/// Returns the decoded `cf_parameter`.
+	fn check_and_decode(
+		ccm: &CcmChannelMetadata,
+		egress_asset: Asset,
+	) -> Result<DecodedCfParameters, CcmValidityError> {
+		if ForeignChain::from(egress_asset) == ForeignChain::Solana {
+			// Check if the cf_parameter can be decoded
+			let ccm_accounts = SolCcmAccounts::decode(&mut &ccm.cf_parameters.clone()[..])
+				.map_err(|_| CcmValidityError::CannotDecodeCfParameters)?;
```

Comment:
We can choose whichever encoding we want, as those will be arbitrary bytes regardless of where they come from (deposit channel or contract swap).
I assumed that scale-encoding was what is more convenient for the State Chain but I must say it was a bit annoying to do in typescript. I tried using the parity `scale-coded` package and hit some issues (can't recall what) and I ended up manually encoding it. I didn't spend much time tbh.
In any case, I'd say fine unless you have a better suggestion.

Reply:
Scale encoding is convenient for us, but if we expect external parties to be able to support this and be able to pass arbitrary data, I would suggest a more universal format. Or for example we might decide that it's based on the native encoding of the external chain's accounts (does that make sense?) and it's the engine's responsibility to transcode this to scale. 

I don't it matters *that* much as long as we are clear about it and make a deliberate choice.

Reply:
The reason why maybe I wouldn't encode it according to the external chain is that it's still a bit unclear what we will use the `cf_parameters` for (besides this use) and it might have uses unrelated to the external destination chain.
Maybe some more universal format would make more sense, I'd just prefer something not very verbose since in contract swaps that will be on-chain data.
Whatever can work better and your end and that is easy to use for integrators/SDK. Let me know what would be a good option and we will have to also update the bouncer accordingly.

Reply:
For general visibility - @albert-llimos and I discussed this and agreed scale encoding makes sense.